### PR TITLE
feat(onlycommit): Process only new commits and generate other formats

### PR DESCRIPTION
**Pull request by heliocastro**
Process only the modified SPDX files and if valid, generate counterparts yaml and json

### CODE REVIEW

**Comment by heliocastro** [comment]
@OliverFendt This commit merge @maxhbr and @sschuberth ideas, and will do two things:
- Process a list of files chnged ( only spdx )
- Generate the counterparts output in yaml and json

@OliverFendt for github actions bot be able to commit on our behalf, you will need to go the project **settings -> Actions -> General -> Workflow permissions** and allow the read/write for the github actions bot

File: .github/workflows/scripts/spdx-tools-java-wrapper.sh [hunk 1]
```diff
@@ -40,18 +41,35 @@ verify() {
...
-    else
-        echo "### All \`.spdx\` files are valid :heavy_check_mark:" > "$GITHUB_STEP_SUMMARY"
+    for package in "$@"; do
+        echo "Analysing package $package"
+        if $javabin -jar "$jar" Verify $package; then
+            # Package is valid. Generate conversion
+            convert "$package"
+        else
+            echo -e "### The following $package are invalid"
+            exit 1
```
Comment (maxhbr) [inline .github/workflows/scripts/spdx-tools-java-wrapper.sh:51]: this fails on first invalid file?

**Comment by maxhbr** [comment]
A few comments

* Even if this checks all newly added `*.spdx` files, it should still check all `*.spdx` files on master (e.g. post merge)
* this is prone to get out of sync. E.g. if just the `*.spdx` gets updated, the `*.json` becomes stale and outdated.

**Comment by heliocastro** [comment]
> 

I don't think so for the regular rule, but a sanitize cron tabbed rule yes, like every week do a checking on the files, or every nightly, just to assure that files are validated

**Review by OliverFendt — COMMENTED** [review]
@heliocastro thank you very much. As far as I understood it, it checks the changed and/or new spdx tag value files and in case there are new and/or modified ones it runs the validation and in case this is passed the other formats are generated. But what I did not see from reading it, are the names of the generated files. For consistency with the TV I thought, the naming convention would be fine if we name those files: packagename-version-JSON.json and packagename-version-YAML-yaml. Any comments on that?

**Comment by OliverFendt** [comment]
Regarding my suggestion on the nameing please see https://github.com/Open-Source-Compliance/package-analysis/tree/main/analysed-packages/libbsd/version-0.11.7 as a reference

**Comment by heliocastro** [comment]
@OliverFendt Basically what is happening on the current PR is in this lines on the script:
```bash
convert() {
    spdx="$1"

    json="${spdx/spdx/json}"
    yaml="${spdx/spdx/yaml}"
```
Is a simple bash replacement for the file extension.
I will update the PR to match your suggestion in a few minutes

File: .github/workflows/scripts/spdx-tools-java-wrapper.sh (hunk 1 above)
Comment (maxhbr) [reply .github/workflows/scripts/spdx-tools-java-wrapper.sh:51]: previously it was "collect all failures, report them, and fail then"

Comment (heliocastro) [reply .github/workflows/scripts/spdx-tools-java-wrapper.sh:51]: Yes, i know. but is just standard CI workflow. If you think we should report any fails, we can restore this behaviour.
But my logic is:

If you build a C++ project, and one of the files fail to build, everything stops. It not continues compiling untils catch everything.

I honestly think in fail soon, catch soon.

**Comment by heliocastro** [comment]
@OliverFendt Pushed new version

I asked myself if we didn't have an official recommendation, and guess what, it has on the SPDX spec:
https://spdx.github.io/spdx-spec/v2.3/conformance/#44-standard-data-format-requirements

So i followed the recommendation, then we are in the safe side.

In the end we will have:
* package-0.0.1-SPDX2TV.spdx
* package-0.0.1.spdx.json
* package-0.0.1.spdx.yaml

Good enough ?

**Comment by OliverFendt** [comment]
@heliocastro cool, that you figured this out. Let's go with that.

**Review by maxhbr — CHANGES_REQUESTED** [review]
I think this is not the best approach for conversion.
In my opinion such a job should run post-merge and for every `.spdx` where no other formats exists **or which was changed after generation of other formats** it should regenerate the other formats. This would prevent race conditions and ensure consistency. It also helps to generate files where the conversion might have failed previously.

File: .github/workflows/scripts/spdx-tools-java-wrapper.sh [hunk 2]
```diff
@@ -40,18 +41,34 @@ verify() {
...
+    done
+}
+
+convert() {
+    spdx="$1"
+
+    json="${spdx/-SPDX2TV/}.json"
+    yaml="${spdx/-SPDX2TV/}.yaml"
+    [ ! -f "$json" ] && rm -f "$json"
+    java -jar "$jar" Convert "$spdx" "$json"
```
Comment (maxhbr) [inline .github/workflows/scripts/spdx-tools-java-wrapper.sh:62]: you are not using `$javabin` here. This is inconsistent

**Comment by maxhbr** [comment]
Hey Helio, I took a small section from your code and adopted #24 to 
* on PR only verify the touched .spdx files
* post merge verify all .spdx files

It also includes some further improvements but does not yet work on the conversion.

I would suggest to merge #24 first and rebase this PR.

**Comment by OliverFendt** [comment]
sounds good, I will merge #24

**Comment by heliocastro** [comment]
@maxhbr @OliverFendt Now i need rewrite my entire commit. 
@maxhbr Do you really need change everything before we finish this one ?

**Comment by maxhbr** [comment]
> Now i need rewrite my entire commit.
> Do you really need change everything before we finish this one ?
my PR was older, so I could ask the same question.

**Comment by maxhbr** [comment]
@heliocastro : maybe we can collaborate and work together, to lift your changes onto the new state of main?

**Comment by OliverFendt** [comment]
@heliocastro and @maxhbr I am sorry that you have some extra work, this was for sure not my intention. I hope that this does not cause any distortion.

**Comment by maxhbr** [comment]
> I think this is not the best approach for conversion. In my opinion such a job should run post-merge and for every `.spdx` where no other formats exists **or which was changed after generation of other formats** it should regenerate the other formats. This would prevent race conditions and ensure consistency. It also helps to generate files where the conversion might have failed previously.

I still think, that this approach for doing conversion will not be stable and consistent. I would like to build a counter-proposal, based on your code here.

File: .github/workflows/scripts/spdx-tools-java-wrapper.sh [hunk 3]
```diff
@@ -89,4 +93,21 @@ verify() {
...
+    spdx="$1"
+
+    local json
+    json="${spdx/-SPDX2TV/}.json"
+
+    local yaml
+    yaml="${spdx/-SPDX2TV/}.yaml"
+
+    for filetype in $json $yaml; do
+        [ -f "$filetype" ] && rm -f "$json"
```
Comment (maxhbr) [inline .github/workflows/scripts/spdx-tools-java-wrapper.sh:107]:
```suggestion
        [ -f "$filetype" ] && rm -f "$filetype"
```

File: .github/workflows/scripts/spdx-tools-java-wrapper.sh [hunk 4]
```diff
@@ -89,4 +93,21 @@ verify() {
...
+        spx_tools Convert "$spdx" "$filetype" 1>&2;
```
Comment (maxhbr) [inline .github/workflows/scripts/spdx-tools-java-wrapper.sh:108]:
```suggestion
        spdx_tools Convert "$spdx" "$filetype" 1>&2;
```

**Comment by heliocastro** [comment]
@maxhbr Here's the rebased code on top of yours.
@OliverFendt Can we merge this now ?

**Review by maxhbr — CHANGES_REQUESTED** [review]

File: .github/workflows/scripts/spdx-tools-java-wrapper.sh [hunk 5]
```diff
@@ -44,7 +44,11 @@ verify_one() {
     GITHUB_REPOSITORY="${GITHUB_REPOSITORY:-.}"
     GITHUB_SHA="${GITHUB_SHA:-..}"
 
-    spdx_tools Verify "$spdx" 1>&2 || echo "* [$spdx]($GITHUB_SERVER_URL/$GITHUB_REPOSITORY/blob/$GITHUB_SHA/$spdx)"
+    if spdx_tools Verify "$spdx" 1>&2; then
+        convert "$spdx"
```
Comment (maxhbr) [inline .github/workflows/scripts/spdx-tools-java-wrapper.sh:48]: If convert writes anything to stdout, this will mess up the other logic.

**Comment by heliocastro** [comment]
@maxhbr Now both cases handle same log output for your logic

**Comment by maxhbr** [comment]
this does not solve the issue. Stdout from convert still finds it way into the `GITHUB_STEP_SUMMARY` and thus would cause the build to be reported as red if there is any.

**Comment by maxhbr** [comment]
> I would like to build a counter-proposal, based on your code here.

@heliocastro , can you have a look at https://github.com/Open-Source-Compliance/package-analysis/pull/29

**Comment by heliocastro** [comment]
@maxhbr #29 can work, but still, i think it should not be complicated and should fail on first error, like any normal development process.

People should do the check *before* commit, in a pre-commit solution, not create problems that complicate pipelines.

@OliverFendt please go with #29 and close this discussion for now

**Pull request closed without merging**